src/java/org/apache/cassandra/index/sai/disk/v1/sortedterms/SortedTermsWriter.java

```diff
@@ -201,14 +234,89 @@ private void copyBytes(ByteComparable source, BytesRefBuilder dest)
             dest.append((byte) val);
     }
 
-    /**
-     * Swaps {@link #tempTerm} with {@link #prevTerm}.
-     * It is faster to swap the pointers instead of copying the data.
-     */
-    private void swapTempWithPrevious()
+    private class TrieSegment
     {
-        BytesRefBuilder temp = this.tempTerm;
-        this.tempTerm = this.prevTerm;
-        this.prevTerm = temp;
+        private final InMemoryTrie<Long> trie;
+        private final BytesRefBuilder prevTerm = new BytesRefBuilder();
+        private final BytesRefBuilder tempTerm = new BytesRefBuilder();
+
+        private BytesRef minTerm;
+        private long totalBytesAllocated;
+        private boolean flushed = false;
+        private boolean active = true;
+
+        TrieSegment()
+        {
+            trie = new InMemoryTrie<>(TrieMemtable.BUFFER_TYPE);
```

Comment:
If we use `OFF_HEAP`, does the limiter just ignore that memory usage?

Reply:
No, only the payloads are stored off-heap, so we still have to report the on-heap memory usage of the trie nodes.